### PR TITLE
Stream over stdin

### PR DESCRIPTION
**Pull request by booniepepper**
IDK how to write a great test for this but:

```
❯ echo -e '1\n2.0\n3.4\n5.6.7' | ./bump
2
2.1
3.5
5.6.8
```

### CODE REVIEW

File: bump.go [hunk 1]
```diff
@@ -20,24 +20,38 @@ type version struct {
...
+
+	for {
+		err := run(reader)
+		
+		if err == io.EOF {
+			os.Exit(0)
+		} else if err != nil {
+			fmt.Fprintf(os.Stderr, "%s\n", err)
+			os.Exit(1)
+		}
```
Comment (lesiw) [inline bump.go]: I don't want to extend `main()` to add this; `main()`'s only job is to check if `run()` errored and to set the return code appropriately / write to standard error.

It's better to extend `run()` and return an error if something failed.

File: bump.go [hunk 2]
```diff
@@ -75,9 +89,8 @@ func parseSegment(s string) (int, error) {
 	return ret, nil
 }
 
-func readInput(reader io.Reader) (string, error) {
-	r := bufio.NewReader(reader)
-	input, err := r.ReadString('\n')
+func readInput(reader *bufio.Reader) (string, error) {
```
Comment (lesiw) [inline bump.go]: I think the right thing to do here would be to change this function to emit `<-chan string` and range over the channel in `run()`.

**Review by lesiw — CHANGES_REQUESTED** [review]

**Comment by lesiw** [comment]
Since we'll be streaming, it's worth thinking about changing how errors will be handled. Right now, we fail and quit the program on unrecognizable input, so `1.2.3\nfoo\n1.0.0\n` would stop after hitting `foo`. I don't think that's the right behavior.

We could probably borrow from `grep` here and no longer consider a failure to match a reason to stop streaming. Instead, parse errors should be written to stderr as they're encountered, but not stop the program.

Also, like `grep`, we might not want to use the return code as an indicator that an error happened, but instead as an indicator whether a successful parse occurred: `0` if at least one parseable version was found in the stream, `1` if not, `2` or greater for a non-streaming error (like flag validation, if we had it; compare to passing `grep` an invalid regex).

**Comment by booniepepper** [comment]
A little rushed but it does the thing

```
❯ echo -e '1\nfoo\n7' | ./bump
2
error bumping version: no version segments found in 'foo'
8

❯ echo -e '1\nfoo\n7' | ./bump 2>/dev/null
2
8

❯ echo -e '1\nfoo\n7' | ./bump >/dev/null
error bumping version: no version segments found in 'foo'
```

Let's think it over a bit tho. Needs tests too

File: bump.go [hunk 3]
```diff
@@ -75,13 +84,25 @@ func parseSegment(s string) (int, error) {
...
+func readInput() <-chan ReadLine {
+	lines := make(chan ReadLine)
+
+    go func() {
+        defer close(lines)
+
+        r := bufio.NewReader(os.Stdin)
+        for {
+			input, err := r.ReadString('\n')
+			lines <- ReadLine{strings.TrimSpace(input), err}
```
Comment (booniepepper) [inline bump.go]: go fmt

File: bump.go [hunk 4]
```diff
@@ -20,10 +20,12 @@ type version struct {
 
 type versionParser func(*version, *bufio.Reader) (versionParser, error)
 
+var segstr *string
```
Comment (booniepepper) [inline bump.go]: move

File: bump.go [hunk 5]
```diff
@@ -36,17 +38,24 @@ func run() error {
...
 
-	output, err := bumpVersion(input, seg)
-	if err != nil {
-		return err
-	}
+	for read := range readCh {
+		if read.Error == io.EOF {
+			return nil
+		} else if read.Error != nil {
+			return read.Error
```
Comment (booniepepper) [inline bump.go]: Should a failed read be an `os.Exit(...)` case?

**Comment by booniepepper** [comment]
YAGNI?

**Pull request merged**